[fio extras]: allow updating meta if the same ver

mike-sul opened the pull request:
Allow updating metadata if local and remote versions are the same but
they actually differ.

Signed-off-by: Mike Sul <mike.sul@foundries.io>

mike-sul committed:
[fio extras]: allow updating meta if the same ver

Allow updating metadata if local and remote versions are the same but
they actually differ.

Signed-off-by: Mike Sul <mike.sul@foundries.io>

vkhoroz commented on src/libaktualizr/uptane/imagerepository.cc:
```diff
@@ -44,7 +45,7 @@ void ImageRepository::fetchSnapshot(INvStorage& storage, const IMetadataFetcher&
   if (local_version > remote_version) {
     throw Uptane::SecurityException(RepositoryType::IMAGE, "Rollback attempt");
-  } else if (local_version < remote_version) {
+  } else if (local_version < remote_version || snapshot_stored_signature != snapshot.signature()) {
```
This seems slightly more complex than `} else {`, but I guess there is reason behind it.
Are you trying to save the amount of database calls ?  Is that performance optimization worth the extra complexity ?

IMHO that totally makes sense for the timestamp role in `updateMeta` as we fetch it quite often.
But I'm not sure about the other roles, as they're only fetched when we've noticed some changes.

Note that in theory equal signature not necessarily means equal content... there is a marginally low chance to get a signature collission.

mike-sul replied:
>Are you trying to save the amount of database calls ? Is that performance optimization worth the extra complexity ?

I am just trying to mimic the logic of `local_version < remote_version` so the business logic for `local_version < remote_version` and `local_version = remote_version (but differ sign)`` is the same.

vkhoroz replied:
sounds reasonable

mike-sul replied:
@pattivacek I agree with @vkhoroz on the condition for storing of snapshot and targets. It can be as simple as just `} else { ` which means store snapshot/targets whenever `local_version <= remote_version` since it's called only in the case if timestamp metadata are updated, and if they are updated it indicates that snapshot/targets are changed so they should be stored in DB. 
I can imagine the case when timestamp are changed while snapshot/targets are not (e.g. timestamp role key rotation) but that's OK to rewrite the metadata in this case since it happens really rarely (and never in the case of Fio).

vkhoroz commented on src/libaktualizr/uptane/imagerepository.cc:
```diff
@@ -219,11 +221,12 @@ void ImageRepository::updateMeta(INvStorage& storage, const IMetadataFetcher& fe
     if (local_version > remote_version) {
       throw Uptane::SecurityException(RepositoryType::IMAGE, "Rollback attempt");
-    } else if (local_version < remote_version) {
+    } else if (local_version < remote_version || timestamp_stored_signature != timestamp.signature()) {
```
One more thing I'm thinking of is this:
- the `verifyXXX` functions store the new data into an instance variable, which can be later reused for different tasks.
- but in case when we throw the security exception above - shouldn't we revert that instance value to the previous value ?

I don't know what other impact it may have, it just seems insane to keep a compomised TUF role in the object state for undefined time.

mike-sul replied:
I know, but effectively that's how it has been  for ages and covered with many tests :). Changing this logic is a bit tricky and it might impact many subtle use-cases.

vkhoroz replied:
yes, you're right

pattivacek replied:
> I don't know what other impact it may have, it just seems insane to keep a compomised TUF role in the object state for undefined time.

I gotta say, I completely agree. This has been something I've been aware of for at least a couple years now. I think the `verifyXXX` functions should be const.

> Changing this logic is a bit tricky and it might impact many subtle use-cases.

I agree that it it's tricky and may have some unexpected subtleties, but I think it's worth doing right. I suspect we're more likely to uncover bugs than break things. And if we break something that relied on a shaky foundation, better to fix it anyway and do it right.

vkhoroz replied:
Maybe, a good trade-off would be really to leave this alone and only do the hot-patch which helps to tackle the exact problem we've seen; and later return back to this when the time allows.

pattivacek replied:
That makes sense for now.

vkhoroz commented:
LGTM

Thanks @mike-sul , it seems you've covered even more use cases than we discussed internally.

pattivacek commented on src/libaktualizr/uptane/tuf.h:
```diff
@@ -153,6 +153,7 @@ class BaseMeta {
   TimeStamp expiry() const { return expiry_; }
   bool isExpired(const TimeStamp &now) const { return expiry_.IsExpiredAt(now); }
   Json::Value original() const { return original_object_; }
+  std::string signature() const { return original_object_["signatures"][0]["sig"].asString(); }
```
This makes me just a bit uncomfortable, because there's a lot of assumptions packed in this one line, and it can throw really easily. At least a `isMember("signatures")` check would help.

mike-sul replied:
I think the assumptions are verified during signature verification. But, then we make an assumption that a signature verification happens before this call (which is true in our case).
Should I just throw exception if the check of "signatures"/0/"sig" fails? I think so since if they are missing then metadata is invalid by definition.
The only concern I have is that timestamp can have more than one signatures and the first signature is the same for both versions while the rest different, but I don't think that's possible as long as the content/subject is the same.

pattivacek replied:
> then we make an assumption that a signature verification happens before this call

I think a decent compromise for now is just to state that assumption as a comment.

> Should I just throw exception if the check of "signatures"/0/"sig" fails? I think so since if they are missing then metadata is invalid by definition.

That wouldn't hurt, either.

> The only concern I have is that timestamp can have more than one signatures and the first signature is the same for both versions while the rest different, but I don't think that's possible as long as the content/subject is the same.

You could also have multiple signatures with different algorithms. We experimented with that once upon a time, but it never went anywhere. I think the single-signature assumption should also be stated in a comment.

vkhoroz replied:
> The only concern I have is that timestamp can have more than one signatures and the first signature is the same for both versions while the rest different, but I don't think that's possible as long as the content/subject is the same.

We do have 2 signatures for production targets at Foundries.io.  So it can be easily imagined that somewhere in a different place or time there might be two signature for the timestamp as well. At one point I considered doing that, so might return back to this topic in the future.

That said, an assumption that there is just one signature seems like a timebomb.  Nothing stops the implementation from shuffling these signatures' order.  Although in the worst case scenario this would just report difference whereas there is equality, in theory that could lead to subtle bugs, which are hard to reproduce.

From this perspective it's safer to always write the new role into the database and let those bugs uncover, just like suggested above.  A more conservative alternative would be to write a function which checks two lists of signatures for equality.

vkhoroz replied:
@mike-sul I've checked in our production targets implementation that we use the consistent order of signatures i.e. the signature for the same key always goes at the same position in the list.  So this patch should work as is.

But we should definitely update this later on with a more consistent solution.
IIRC we've had a long discussion a year ago about why HAVING consistent order for signatures is a good thing while RELYING onto this consistent order is a bad thing.

mike-sul replied:
> IIRC we've had a long discussion a year ago about why HAVING consistent order for signatures is a good thing while RELYING onto this consistent order is a bad thing.

If signature order is different in the same versions of different metadata then we are fine because  stored_sign != update_sign and we store the new version of metadata.

If signature order is different in the same versions of the same metadata then we are fine too because  stored_sign != update_sign and we just store the same metadata again.

pattivacek replied:
> IIRC we've had a long discussion a year ago about why HAVING consistent order for signatures is a good thing while RELYING onto this consistent order is a bad thing.

Fully agree.

> If signature order is different in the same versions of the same metadata then we are fine too because stored_sign != update_sign and we just store the same metadata again.

That is a fair point. Not super efficient, but probably fine. My thought is that there probably isn't a lot of need to worry about this now, but if, down the line, there are more complexities around the signatures, this would be worth revisiting.

pattivacek reviewed (commented):
This is great, thank you. I think this needs some testing, and I'd love it if we can constify the `verifyXXX` functions, but maybe that bit can come later.

mike-sul commented:
> if we can constify the `verifyXXX` functions, but maybe that bit can come later.

I think it makes sense, we just need to carve out time slot to get it done :)

mike-sul commented:
@pattivacek I updated the PR with the suggested improvements. 
I tend to agree with @vkhoroz in terms of considering it as kind of hotfix, since the assumption about one signature can potentially lead to very subtle issue(s). Since it's gonna take some time to do it really properly (hashes, and we don't have timestamp hash) I suggest to merge this and create ticket so we eventually implement metadata comparison absolutely correct (although in theory hash collisions may happen :) ).

mike-sul committed:
[fio extras]: add 'length' and 'hashes' to snapshot

Add 'length' and 'hashes' attributes of targets to snapshot metadata.
It is required for a case when different targets have the same version.
If just a version attribute is specified then snapshots of these two
different targets are identical, hence it's not possible for a client
to determine whether to update metadata or not.

Signed-off-by: Mike Sul <mike.sul@foundries.io>

vkhoroz commented on src/uptane_generator/repo.cc:
```diff
@@ -39,6 +39,10 @@ void Repo::addDelegationToSnapshot(Json::Value *snapshot, const Uptane::Role &ro
+  (*snapshot)["meta"][role_file_name]["length"] = signed_role.size();
+  (*snapshot)["meta"][role_file_name]["length"] = signed_role.size();
+  (*snapshot)["meta"][role_file_name]["hashes"]["sha256"] =
+      boost::algorithm::to_lower_copy(boost::algorithm::hex(Crypto::sha256digest(signed_role)));
```
Just two questions:
1. why do we set length twice ?
2. how does it help with the fix ?

mike-sul replied:
1. just typo
2. it doesn't help with the fix, it's needed for the test.

vkhoroz commented:
LGTM

pattivacek commented on src/libaktualizr/uptane/imagerepository.cc:
```diff
@@ -44,7 +44,7 @@ void ImageRepository::fetchSnapshot(INvStorage& storage, const IMetadataFetcher&
   if (local_version > remote_version) {
     throw Uptane::SecurityException(RepositoryType::IMAGE, "Rollback attempt");
-  } else if (local_version < remote_version) {
+  } else {
```
I preferred the logic where we compared signatures, because this means that we will store the same thing repeatedly if it hasn't changed. That causes unnecessary writes to the database that I'd like to avoid.

pattivacek replied:
It seems like the Timestamp is being treated differently than Snapshot and Targets. I can imagine there's good reason for it, but it's not clear to me. What's the reasoning there?

mike-sul replied:
The additional condition `|| timestamp_stored_signature != timestamp.signature()` helps to avoid rewriting the same timestamp metadata in DB each time timestamp metadata are downloaded and checked at the beginning of each update cycle.
This additional condition doesn't bring much benefit in the case of snapshot and targets metadata because these metadata are downloaded only if timestamp metadata are changed (verifySnapshot()/verifyTargets() throws exception, otherwise `fetch_snapshot/fetch_targets` are set to false).
This is very unlikely or can happen very rarely when timestamp changes while snapshot&targets are not, therefore the given approach, helps to avoid unnecessary writes to DB in majority cases.

pattivacek replied:
Yes, you make a good point. I do kinda wish these things were made more clear in the code or with comments, though. And I still think low-level tests are important.
